### PR TITLE
Do not set exception_check if already set during analyzing pxd file

### LINKED ISSUE #6122: [BUG] 'except?' started failing in 3.0.10
### Describe the bug

Usage of `except? -1` has seemingly started failing in 3.0.10. It works in all previous >3.0.0 versions.

Every place I use the "?"-version of `except`, it fails during `cythonize` with:
```
pyboy/utils.pxd:22:29: C method 'read_64bit' is declared but not defined
```

Note, it says "C method", but this is purely a Python/Cython function, it's not defining an external library.

### Code to reproduce the behaviour:

I define a `e.pxd` file with:
```cython
from libc.stdint cimport uint64_t

cdef class IntIOInterface:
    cpdef uint64_t read_64bit(self) except? -1
```

And a `e.py` file with:
```python
class IntIOInterface:

    def read_64bit(self):
        a = self.read()
        return a
```

Compile using following command:
```
cython --directive legacy_implicit_noexcept=True -3 e.py
```
### Expected behaviour

Expected to compile cleanly, and use `-1` to signal an exception like in previous versions.

### OS

_No response_

### Python version

_No response_

### Cython version

3.0.10

### Additional context

The code can be found here: https://github.com/Baekalfen/PyBoy/blob/ef25b6bb6445b6228069df57297620b0fe9487d8/pyboy/utils.pxd#L17

This bug report and fix is listed for Cython 3.0.10: https://github.com/cython/cython/issues/5709

### PR DESCRIPTION
**Pull request by matusvalo**
When .pxd file is defined there are two runs CFuncDeclaratorNode.analyse:

1. run is executed for .pxd file and is triggered by the parser
2. run is executed for .py file and is triggered by ParseTreeTransform.AlignFunctionDefinitions.visit_DefNode

The problem is causing 2nd run because it does uses information from entry which misses correct value of has_explicit_exc_clause (default value False is used).

PR solves it by not overriding exception_check when CFuncDeclaratorNode is already declared by .pxd file since it already contains correct value of exception_check from 1st run. 

Fixes #6122

### CODE REVIEW

File: tests/run/legacy_implicit_noexcept_build.srctree [hunk 1]
```diff
@@ -31,3 +32,16 @@ except RuntimeError:
     pass
 else:
     assert False, 'Exception not raised'
+
+
+######## aa.pxd ########
+cdef class Klass:
+    cpdef int m(self) except? -1
+
```
Comment (da-woods) [inline tests/run/legacy_implicit_noexcept_build.srctree:40]:
```suggestion
    cpdef int m2(self)
    cpdef int m3(self) noexcept
    cpdef int m4(self) except *
    cpdef int m5(self) except -1
    cpdef object n1(self)
    cpdef object n2(self) except *
    cpdef object n3(self) noexcept  # will warn, but should work
    cpdef void o1(self)
    cpdef void o2(self) except*
    cpdef void o3(self) noexcept
    cdef int p1(self) except? -1
    cdef int p2(self)
    cdef int p3(self) noexcept
    cdef int p4(self) except *
    cdef int p5(self) except -1
    cdef object q1(self)
    cdef object q2(self) except *
    cdef object q3(self) noexcept  # will warn, but should work
    cdef void r1(self)
    cdef void r2(self) except*
    cdef void r3(self) noexcept

cpdef int m1(self) except? -1
cpdef int m2()
cpdef int m3() noexcept
cpdef int m4() except *
cpdef int m5() except -1
cpdef object n1()
cpdef object n2() except *
cpdef object n3() noexcept  # will warn, but should work
cpdef void o1()
cpdef void o2() except*
cpdef void o3() noexcept
cdef int p1() except? -1
cdef int p2()
cdef int p3() noexcept
cdef int p4() except *
cdef int p5() except -1
cdef object q1()
cdef object q2() except *
cdef object q3() noexcept  # will warn, but should work
cdef void r1()
cdef void r2() except*
cdef void r3() noexcept
```

Since this is obviously a little under-tested, just just suggesting adding a long list of similar examples here in the hope that it catches any other edge cases

Comment (matusvalo) [reply tests/run/legacy_implicit_noexcept_build.srctree]: Fixed. I think I should also implement tests for exception propagation (non-propagation of noexcept functions) for every method/function I suppose...

Comment (matusvalo) [reply tests/run/legacy_implicit_noexcept_build.srctree:228]: >  think I should also implement tests for exception propagation

done

**Comment by matusvalo** [comment]
OK I think PR is ready for merge. Let me know if something needs to be changed.

**Review by da-woods — APPROVED** [review]
I think this is good.

I'd like to check it against [this issue from the mailing list](https://groups.google.com/g/cython-users/c/q5FzO2d88O) but I think that's a follow up and doesn't need to be in this PR.

**Comment by matusvalo** [comment]
> I'd like to check it against [this issue from the mailing list](https://groups.google.com/g/cython-users/c/q5FzO2d88Og)

If I am not wrong, it does not fix the issue unfortunately :-(

**Comment by da-woods** [comment]
> If I am not wrong, it does not fix the issue unfortunately :-(

I'm not very surprised that doesn't solve it. Thanks for checking though.

**Comment by tornaria** [comment]
This breaks a function that is implicit noexcept in the pxd file as in:
```
$ cat a.pxd 
cdef int f()
$ cat a.pyx 
cdef int f():
    pass
$ cython -3X legacy_implicit_noexcept=True a.pyx 
warning: a.pxd:1:10: Implicit noexcept declaration is deprecated. Function declaration should contain 'noexcept' keyword.
warning: a.pyx:1:10: env.entries[self.declared_name()]=Entry(<7f78f7ec0ce0>, name=f, type=int (void) noexcept). 

Error compiling Cython file:
------------------------------------------------------------
...
cdef int f():
     ^
------------------------------------------------------------

a.pyx:1:5: Function signature does not match previous declaration
```

In particular, sagemath fails to build with cython 3.0.11.

**Comment by matusvalo** [comment]
> This breaks a function that is implicit noexcept in the pxd file as in

Based on comment in linked bug report here is possible workaround:

```cython
# cython: legacy_implicit_noexcept=False
cdef int f()
```

```cython
```

**Comment by matusvalo** [comment]
It seems that this fix is really broken and should be reverted and fix should be implemented in different way:

It seems that there are two calls of `CFuncDeclaratorNode.analyse` one for .pxd file and other one for .py file.

Both calls differ in value of `self.has_explicit_exc_clause` .pxd has correct value `True` and .py has incorrect value `False` (for example given in the #6122). This is wrong and is because following:

CFunc node from def node is created here:
https://github.com/cython/cython/blob/6687185974610d2b9cd87a9cbd33cb94cfd73c09/Cython/Compiler/Nodes.py#L3146-L3168

In the example it can be seen that exception information is created from Entry type `Cython.Compiler.PyrexTypes.CFuncType` which does not contain information whether declaration of exception value is implicit or explicit and **hence all the time the default `False` value is used (regardless of what is in .pxd file)**.

One possible solution is to extend `Cython.Compiler.PyrexTypes.CFuncType` to contain also information whether exception definition is implicit or explicit. Not sure if it is worth it though :-(. I think users should migrate to new semantics if possible.

**Comment by tornaria** [comment]
> In the example, it can be seen that exception information is created from Entry type `Cython.Compiler.PyrexTypes.CFuncType` (entry type is coming from .pxd file definition) which does not contain information whether declaration of exception value is implicit or explicit and **hence all the time the default `False` value is used (regardless of what is in .pxd file)**.

If I understand it correctly, the idea is that the exception value is taken from the `.pxd` and it doesn't need to be present in the `.py` file.

Maybe when the exception value is taken from the `.pxd`, the value of `has_explicit_exc_clause` should always be `True`?

This assumes that when the `.pxd` has an implicit `noexcept`, the exception value will be `noexcept` when processing the `.py` file.

> One possible solution is to extend `Cython.Compiler.PyrexTypes.CFuncType` to contain also information whether exception definition in .pxd file is implicit or explicit and feed this information to `CFuncDeclaratorNode`

Maybe not necessary as suggested above?


> Not sure if it is worth it though :-(. I think users should migrate to new semantics if possible.

Sure, but having `legacy_implicit_noexcept` work ok is necessary for a smooth migration. And the "legacy implicit noexcept" warning should coincide with what is going on in the sense that:
 - if there is a "legacy implicit noexcept" warning, and a `noexcept` clause is added to that line, the warning is suppressed and the behaviour should be identical.
 - once there are no "legacy implicit noexcept" warnings, it should be safe to remove the option, and the behaviour should be identical.

**Comment by tornaria** [comment]
Maybe this is reasonable:
```diff
--- a/Cython/Compiler/Nodes.py
+++ b/Cython/Compiler/Nodes.py
@@ -3128,6 +3126,7 @@ class DefNode(FuncDefNode):
             if scope is None:
                 scope = cfunc.scope
             cfunc_type = cfunc.type
+            has_explicit_exc_clause=True
             if len(self.args) != len(cfunc_type.args) or cfunc_type.has_varargs:
                 error(self.pos, "wrong number of arguments")
                 error(cfunc.pos, "previous declaration here")
```

The idea is that, when `cfunc` is given, we will be using `exception_value` and `exception_check` from the given `cfunc` so we should act as if an explicit exception clause was given.


Don't forget to revert the previous fix as in:
```diff
--- a/Cython/Compiler/Nodes.py
+++ b/Cython/Compiler/Nodes.py
@@ -710,10 +710,8 @@ class CFuncDeclaratorNode(CDeclaratorNode):
                 and not self.has_explicit_exc_clause
                 and self.exception_check
                 and visibility != 'extern'):
-            # If function is already declared from pxd, the exception_check has already correct value.
-            if not (self.declared_name() in env.entries and not in_pxd):
-                self.exception_check = False
             # implicit noexcept, with a warning
+            self.exception_check = False
             warning(self.pos,
                     "Implicit noexcept declaration is deprecated."
                     " Function declaration should contain 'noexcept' keyword.",
```
